### PR TITLE
Allow Kernels for Full FT and Non-Quantized PEFT

### PR DESCRIPTION
**Pull request by fabianlim**
## Description

This PR
1. upgrades framework to perform OR logic when activating plugins
2. creates a `FastKernelsAccelerationPlugin` that is an improved version over `FastQuantizedPeftAccelerationPlugin`
  * it can add kernels individually
  * it can be activated under an `training` stanza or a `peft.quantized` stanza
3. Add FOAK support to Full-Finetuning and Standard PEFT benchmarks
4. FOAK support on 1 additional models
    - GPTBigCode
        - Note that due to GPTBigCode architecture limitations only FastCrossEntropyLoss is supported in this PR. Additional support will be tracked [placeholder issue]
5. Bug fix to ModelPatcher to address multiple reloads to the same target path

### Improvements to Full Finetuning

7% Improvement from following kernels (FastCrossEntropyLoss, FastRMSNorm, FastRoPE)
Framework | Model | num gpus | batch size| throughput (toks/s) | Improvement %
-- | -- | -- | -- | -- | --
fullFT | Mistral7B | 1 | 4 | 2910 | base
foak-fullFT | Mistral7B | 1 | 4 | 3218 | 10.5
PEFT | Mistral7B | 1 | 4 | 3345 | base
foak-PEFT | Mistral7B | 1 | 4 | 3797 | 13.5

Framework | Model | num gpus | batch size| throughput (toks/s) | Improvement %
-- | -- | -- | -- | -- | --
fullFT | Mistral7B | 2 | 4 | 2886 | base
foak-fullFT | Mistral7B | 2 | 4 | 3093 | 7
PEFT | Mistral7B | 2 | 4 | 3227 | base
foak-PEFT | Mistral7B | 2 | 4 | 3620 | 12

<Add Larger Models>

<Add Larger Models>

**Compatibility Matrix with Mixed Precision**
torch_dtype | Mixed Precision | Full-FT-FOAK | PEFT-FOAK | QPEFT-FOAK
-- | -- | -- | -- | --
FLOAT16 | - | ✗ Not Allowed | ✗| ✗
FLOAT16 | FP16 | ValueError: <br>Attempting to <br>unscale FP16 gradients. <br>[See here](https://github.com/huggingface/peft/blob/main/docs/source/developer_guides/troubleshooting.md) | **Compatible** | **Compatible**
BFLOAT16 | - | ✗ | ✗ | ✗
BFLOAT16 | BF16 | **Compatible** | **Compatible** | [Less Performant](https://github.com/foundation-model-stack/fms-acceleration/issues/84)

### Improvements to PEFT and QPEFT

**Note:**
Due to issues with FSDP-QLoRA in the latest transformers version (`4.45.0dev`) mentioned [here](https://github.com/foundation-model-stack/fms-acceleration/issues/83), Granite with Fast Kernels will be addressed in a later PR instead.

### TODO
- add the activation (e.g. SWIGLU) kernels to `FastKernelsAccelerationPlugin`. Follow the pattern of building the fused-lora rule for a `base_type`.
- add chunked loss (optional). If not done create issue.

### CODE REVIEW

File: plugins/fused-ops-and-kernels/configs/fast_full.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+training:
+
+  fused_ops_and_kernels: 
+
+    # if under training stanza, then putting
+    # base_layer and fused_lora will be a misnomer
+    # - this should be in peft.quantized
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/configs/fast_full.yaml:7]: maybe should change the name, this also works for non-quantized peft

File: plugins/fused-ops-and-kernels/src/fms_acceleration_foak/models/granite.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+from functools import partial
+
+# Third Party
+from fms_acceleration.model_patcher import (
+    ModelPatcherRule,
+    ModelPatcherTrigger,
+    combine_functions,
+    combine_triggers,
+)
+from transformers.models.granite.modeling_granite import (
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/models/granite.py:25]: this needs to be gaurded because it only exists for later enough transformer versions
1. move this inside `get_mp_rules`
2. gaurd it inside a try catch
3. if `except`, then return and empty `[]`

File: scripts/generate_sample_configurations.py [hunk 3]
```diff
@@ -171,6 +172,7 @@ def read_configuration(path: str) -> Dict:
     ),
     KEY_AADP_PADDING_FREE: "plugins/attention-and-distributed-packing/configs/padding_free.yaml",
     KEY_AADP_MULTIPACK: "plugins/attention-and-distributed-packing/configs/multipack.yaml",
+    KEY_FAST_KERNELS: "plugins/fused-ops-and-kernels/configs/fast_full.yaml",
```
Comment (fabianlim) [inline scripts/generate_sample_configurations.py:175]: if this one is used for both `peft` and `full`, then we need to name it as such

File: scripts/benchmarks/scenarios.yaml [hunk 4]
```diff
@@ -65,7 +73,7 @@ scenarios:
         arguments:
             fp16: True
```
Comment (fabianlim) [inline scripts/benchmarks/scenarios.yaml:74]: have to switch to `bf16`.

File: plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            if v:
+                ts = FILTER_MAP[k]
+                if isinstance(ts, str):
+                    ts = {ts}
+                terms.update(ts)
+
+        # wrapper function to register foak patches
+        # NOTE: we never take the lora modules so just set arbitrarily
+        # to "auto_gptq"
+        _base_layer = self.configurations['base_layer'] if 'base_layer' \
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py:137]: lets not do this.. its anti pattern, we already have a `default` check above

File: plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        except AccelerationPluginConfigError:
+            self.configurations = self._check_config_and_maybe_check_values(
+                key="peft.quantization.fused_ops_and_kernels",
+            )
+
+        self._check_config_and_maybe_check_values(
+            key="base_layer",
+            values=["auto_gptq", "bitsandbytes"],
+            default="auto_gptq"
+        )
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py:98]:
```suggestion
        self.configurations['base_layer'] = self._check_config_and_maybe_check_values(
            key="base_layer",
            values=["auto_gptq", "bitsandbytes"],
            default="auto_gptq"
        )
```

File: plugins/framework/pyproject.toml [hunk 7]
```diff
@@ -24,7 +24,7 @@ classifiers=[
 dependencies = [
   "numpy<2.0", # numpy needs to be bounded due to incompatiblity with current torch<2.3
   "torch>2.2",
-  "transformers",
+  "git+https://github.com/huggingface/transformers.git@9230d78e76611cfa38c845213021aeb185362d10",
```
Comment (fabianlim) [inline plugins/framework/pyproject.toml:27]: pls revert this change

File: plugins/fused-ops-and-kernels/README.md [hunk 8]
```diff
@@ -37,6 +39,10 @@ Path | Description | Extracted From  | Modifications | Date
 [fused_ops/unsloth_lora/gptq](./src/fms_acceleration_foak/fused_ops/unsloth_lora/gptq) | GPTQ fast dequant (triton_v2) | `jeromeku/main` @ [2839d39](https://github.com/jeromeku/unsloth/commit/2839d390ef3bb318904289bfb9a7751a782c4e44) | `fast_lora.py`<br>`triton/layers.py` | 6 Feb 2024
 [kernels/unsloth](./src/fms_acceleration_foak/kernels/unsloth) | Fast RMS, RoPE, CrossEnt kernels | `unsloth/main` @ [1ecc0185](https://github.com/unslothai/unsloth/commit/1ecc0185a5759c7a0c95dfc96aceea5023cebdfc) | `cross_entropy_loss.py`<br>`rms_layernorm.py` | 28 Jan 2024
 
+
+## Code Extracted from Liger
+- TODO
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/README.md:44]: needs to be updated. eitgher
1. say in the future we will import the fused cross ent
2. delete the block first then we update this when we get to this task

File: plugins/fused-ops-and-kernels/README.md [hunk 9]
```diff
@@ -37,6 +39,10 @@ Path | Description | Extracted From  | Modifications | Date
...
+
 ## Known Issues
 
 - MixedPrecision `--fp16` should be used `fast_lora`. Also consider loading the model in `torch.float16`.
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/README.md:48]: potentially we might need to update all references of `float16`, even those in `fms-hf-tuning`

File: plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            r for r in rules if 
+            any(r.rule_id.endswith(x) for x in filter_endswith)
+        ]
+
+    for _rule in rules:
+        ModelPatcher.register(_rule)
+
+# maybe this we should define envvars
+FILTER_MAP = {
+    "base_layer": set(),
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py:62]: remove this and below, check if `k` is in `FILTER_MAP`

File: scripts/generate_sample_configurations.py (hunk 3 above)
Comment (fabianlim) [reply scripts/generate_sample_configurations.py:175]: maybe something like `fast_kernels.yaml`

File: plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py [hunk 11]
```diff
@@ -275,9 +275,6 @@ def augmentation(
 
         # some assertions
         assert peft_config is not None, "need peft_config to install PEFT adapters"
-        assert (
```
Comment (fabianlim) [inline plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py:278]: ok so, we possibly will need to move this to foak plugin.  my guess is that the only thing we need to enforce is

```python
if model.dtype == torch.float16`:
    assert train_args.fp16
```

Comment (fabianlim) [reply plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py:278]: test

dtype | mp | outcome
--|--|--
float16 | Y |
foat16 | N|
bfloat16 | Y 
bfloat16 | N

**Review by fabianlim — COMMENTED** [review]
When you get the bfloat16 numbers, lets compare with the float16 numbers to see if there are substantial changes. 

Also, lets have in the FoAK readme, to documnet down future items for kernels that are missing for certain models

Model | norm | pos emb | cross-ent | fused_lora 
--|--|--|--|--
`LlamaForCausalLM` | ✅  | ✅ | ✅  | ✅

**Comment by fabianlim** [comment]
@achew010 pls also note that currently we do support pos ids with rope kernels. we need to document the impact on this

https://github.com/foundation-model-stack/fms-acceleration/issues/33

I think if its padding free its no impact, but need to confirm

File: plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py [hunk 12]
```diff
@@ -230,8 +230,9 @@ def model_loader(self, model_name: str, **kwargs):
         # - in particular "is_loaded_in_4bit" will be checked in prepare_model_for_kbit_training
         #   and there is a section of code that will be skipped if not set.
         setattr(model, "is_loaded_in_4bit", True)
-        setattr(model, "quantization_method", "gptq")
-
+        # Need to set this on model.model instead of model, applies to all versionsof gptq
+        # as this attribute is accessed later on from model.model
+        setattr(model.model, "quantization_method", "gptq")
```
Comment (fabianlim) [inline plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py:235]: dont change this here, copy it out a the last moment where it disappears.

File: scripts/benchmarks/scenarios.yaml [hunk 13]
```diff
@@ -37,77 +37,87 @@
...
             r: 16
             lora_alpha: 16
             lora_dropout: 0.1
-            target_modules: ["q_proj", "k_proj", "v_proj", "o_proj"]
+            target_modules: ["q_proj", "k_proj", "v_proj", "o_proj", "c_attn"]
             model_name_or_path: 
+                # - 'ibm/PowerLM-3b'
+                - 'bigcode/gpt_bigcode-santacoder'
                 - 'mistralai/Mistral-7B-v0.1'
-                - 'mistralai/Mixtral-8x7B-Instruct-v0.1'
```
Comment (fabianlim) [inline scripts/benchmarks/scenarios.yaml:94]: these comments need to be reverted.

File: plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py [hunk 14]
```diff
@@ -321,8 +321,7 @@ def augmentation(
             train_mode=True,  # install adapaters for training
         )
 
-        # Reassign the attributes after PEFT installation replaces the top-level class
-        setattr(model, "is_loaded_in_4bit", True)
+        # Reassign `quantization_method` after PEFT installation replaces the top-level class
```
Comment (fabianlim) [inline plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py:324]: I thought that it will also be the case that `is_loaded_in_4bit` be also dropped. Then can for consistency we should just set it back also

File: plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_quantized_peft.py [hunk 15]
```diff
@@ -160,9 +160,9 @@ def get_callbacks_and_ready_for_train(
             )
         return callbacks
 
-
-# register
-AccelerationPlugin.register_plugin(
-    FastQuantizedPeftAccelerationPlugin,
-    configuration_and_paths=["peft.quantization.fused_ops_and_kernels"],
-)
+# This plugin is currently deregistered in favour of fast_kernels.py
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_quantized_peft.py:163]:
```suggestion
# This plugin is currently deregistered in favour of framework_plugin_fast_kernels.py
```

File: plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+                    ts = {ts}
+                terms.update(ts)
+
+        # wrapper function to register foak patches
+        # NOTE: we never take the lora modules so just set arbitrarily
+        # to "auto_gptq"
+        _base_layer = self.configurations.get('base_layer', None)
+        register_foak_model_patch_rules2(
+            base_type=_base_layer, filter_endswith=terms
+        )
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py:145]:
```suggestion
        register_foak_model_patch_rules2(
            base_type=self.configurations['base_layer'], 
            filter_endswith=terms
        )
```

Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py:145]:
```suggestion
        # wrapper function to register foak patches
        # - the base layer setting below will be ignored in non quantized-lora settings
        register_foak_model_patch_rules2(
            base_type=self.configurations['base_layer'], 
            filter_endswith=terms
        )
```

File: plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+        
+        terms = set()
+        for k in self.configurations:
+            if k in FILTER_MAP:
+                if k == "fused_lora" and not is_quantized:
+                    continue
+                ts = FILTER_MAP[k]
+                if isinstance(ts, str):
+                    ts = {ts}
+                terms.update(ts)
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/framework_plugin_fast_kernels.py:137]:
```suggestion

        # some logic to omit terms from the filter if logic precludes
        omitted = set()
        if getattr(model, "quantization_method", None):
            ommited.add("fused_lora")
        
        terms = set()
        for k in self.configurations:
            if k in FILTER_MAP and k not in omitted:
                ts = FILTER_MAP[k]
                if isinstance(ts, str):
                    ts = {ts}
                terms.update(ts)
```

File: plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py (hunk 14 above)
Comment (achew010) [reply plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py:324]: Setting the `is_loaded_in_4bit` attribute will cause an error in `accelerator.prepare_model` just before the FSDP wrapping. FSDP will eventually handle the device placements later on so not sure if this attribute is useful in the GPTQ-LoRA case.

```
  File "/data/aaron/experimental/fms-hf-tuning/tuning/sft_trainer.py", line 360, in train
    trainer.train()
  File "/data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/trl/trainer/sft_trainer.py", line 450, in train
    output = super().train(*args, **kwargs)
  File "/data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/transformers/trainer.py", line 1938, in train
    return inner_training_loop(
  File "/data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/transformers/trainer.py", line 2085, in _inner_training_loop
    self.model = self.accelerator.prepare(self.model)
  File "/data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/accelerate/accelerator.py", line 1311, in prepare
    result = tuple(
  File "/data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/accelerate/accelerator.py", line 1312, in <genexpr>
    self._prepare_one(obj, first_pass=True, device_placement=d) for obj, d in zip(args, device_placement)
  File "/data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/accelerate/accelerator.py", line 1188, in _prepare_one
    return self.prepare_model(obj, device_placement=device_placement)
  File "/data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/accelerate/accelerator.py", line 1425, in prepare_model
    raise ValueError(
ValueError: You can't train a model that has been loaded in 8-bit precision on a different device than the one you're training on. Make sure you loaded the model on the correct device using for example `device_map={'':torch.cuda.current_device()}` or `device_map={'':torch.xpu.current_device()}`
```

Comment (fabianlim) [reply plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py:324]: ok then can you put a comment 

```We do not set is_loaded_in_4bit at this point because otherwise `accelerate.prepare_model` will think the device placement is finalized for the quantized model, and will raise```

File: plugins/fused-ops-and-kernels/README.md [hunk 18]
```diff
@@ -37,11 +39,20 @@ Path | Description | Extracted From  | Modifications | Date
...
+`MistralForCausalLM` | ✅  | ✅ | ✅  | ✅ 
+`MixtralForCausalLM` | ✅  | ✅ | ✅  | ✅ 
+`GPTBigCodeForCausalLM` | ❌  | ❌ | ✅  | ❌ 
+<!-- `GraniteForCausalLM` | ✅  | ✅ | ✅  | ✅  -->
+
 ## Known Issues
 
-- MixedPrecision `--fp16` should be used `fast_lora`. Also consider loading the model in `torch.float16`.
-- `fast_lora` has issues with FSDP with the `peft` style of FSDP wrapping. 
+- MixedPrecision `--fp16` or `--bf16` should be used with `fast_lora`.
```
[the remark this answers is not in the document]
Comment (fabianlim) [reply plugins/fused-ops-and-kernels/README.md:54]: Put it under ### Supported DataType Settings

File: plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py [hunk 19]
```diff
@@ -274,7 +274,9 @@ def augmentation(
 
         # some assertions
         assert peft_config is not None, "need peft_config to install PEFT adapters"
-
+        assert (
```
Comment (fabianlim) [inline plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py:277]: can you put link to this issue in comments https://github.com/foundation-model-stack/fms-acceleration/issues/84

File: scripts/benchmarks/scenarios.yaml [hunk 20]
```diff
@@ -71,9 +73,9 @@ scenarios:
         framework_config: 
             - baseline-peft-bnb
         arguments:
-            bf16: True
+            fp16: True
```
Comment (fabianlim) [inline scripts/benchmarks/scenarios.yaml:76]: bf16 for `baseline-peft-bnb`, `accelerated-peft-bnb`,

File: scripts/benchmarks/scenarios.yaml [hunk 21]
```diff
@@ -110,8 +112,8 @@ scenarios:
             - accelerated-peft-autogptq-foak
         arguments:
             learning_rate: 2e-4
-            bf16: True
-            torch_dtype: bfloat16
+            fp16: True
```
Comment (fabianlim) [inline scripts/benchmarks/scenarios.yaml:115]: you can put even a link to this issue in comments https://github.com/foundation-model-stack/fms-acceleration/issues/84

Comment (fabianlim) [inline scripts/benchmarks/scenarios.yaml:115]: you can put even a link to this issue in comments https://github.com/foundation-model-stack/fms-acceleration/issues/84

**Comment by fabianlim** [comment]
@achew010 there are some benchmarks still not updated, but I will merge this first then we can address in later PR
